DOCSP-9555: indexes

terakilobyte opened the pull request:
## Pull Request Info

Adding coverage for indexes. A link to the builders page will be added as part of that PR.

### Issue JIRA link:
https://jira.mongodb.org/browse/DOCSP-9555

### Docs staging link (requires sign-in on MongoDB Corp SSO):
https://docs-mongodbcom-staging.corp.mongodb.com/1bf9dff/java/docsworker-xlarge/DOCSP-9555/fundamentals/indexes/

### Self-Review Checklist

- [x] Is this free of any warnings or errors in the RST?
- [x] Did you run a spell-check?
- [x] Did you run a grammar-check?
- [x] Does it render on staging correctly?
- [x] Are all the links working?
- [x] Are the staging links in the PR description updated?

nathan-contino-mongo commented on source/fundamentals/indexes.txt:
```diff
@@ -0,0 +1,253 @@
+**Single field indexes** are indexes that improve performance for queries that specify ascending or descending sort
+order on a single field of a document.
+
+The following example uses the ``createIndex()`` method to create an ascending order index on the ``title`` field in the
```
In the Realm docs, we'd usually link to the reference doc for `createIndex()` here. Would that be a good idea here?

nathan-contino-mongo commented on source/fundamentals/indexes.txt:
```diff
@@ -0,0 +1,253 @@
+order on a single field of a document.
+
+The following example uses the ``createIndex()`` method to create an ascending order index on the ``title`` field in the
+``movies`` collection in the ``sample_mflix`` database.
```
This code snippet doesn't necessarily refer to a "movies" collection of a "sample_mflix" database. Should we add some lines to this snippet instantiating those objects, or should we just make this line more generic with something like "...field in a collection"?

nathan-contino-mongo replied:
I noticed that you don't use a colon to segue these snippet descriptions into the snippets themselves. I think most MongoDB documentation uses colons for situations like these -- would it make sense to switch from period to colons?

nathan-contino-mongo commented on source/fundamentals/indexes.txt:
```diff
@@ -0,0 +1,253 @@
+index.
+
+The following example uses the ``createIndex()`` method to create a compound index on the ``type`` and ``rated`` fields
+in the ``movies`` collection in the ``sample_mflix`` database.
```
Same as above -- should we add lines to the code snippet demonstrating instantiating the "movies" collection and "sample_mflix" database, or make this line more generic?

nathan-contino-mongo commented on source/includes/fundamentals/code-snippets/IndexPage.java:
```diff
@@ -0,0 +1,150 @@
+        System.out.println("single index");
+        // begin single index
+        String resultCreateIndex = collection.createIndex(Indexes.ascending("title"));
+        System.out.println(String.format("Index created: %s", resultCreateIndex));
```
Is there a style reason you'd use String.format here instead of printf, or just personal choice? I'd think that printf would be easier to understand for intermediate-level developers.

terakilobyte replied:
I used this to avoid having to do `System.out.printf("Index created: %s%n", resultCreateIndex);` for the newline.

nathan-contino-mongo commented on source/fundamentals/indexes.txt:
```diff
@@ -0,0 +1,253 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+**Multikey indexes** are indexes that improve performance for queries that specify ascending or descending indexes on
+fields that contain an array value. You can define a multikey index using the same syntax as a single field or compound
```
Personal opinion here: I think this sentence gets a bit lost in the weeds of "queries that specify ascending or descending indexes" to a point where it's easy to miss the actual idea (multikey indexes are indexes on fields that contain an array value). Arguably you've mitigated this issue in the title of this subsection, but I wonder if we could trim this sentence a little to make it less wordy.

nathan-contino-mongo commented on source/includes/fundamentals/code-snippets/IndexPage.java:
```diff
@@ -0,0 +1,150 @@
+        // begin text index
+        // create a text index of the "fullplot" field in the "movies" collection
+        // if a text index already exists with a different configuration, this will
+        // error
```
This comment brings up a pretty important footgun w.r.t. text indexes -- the fact that you can only have one per collection. The rest of the comment is rendered redundant by the description. Would it make sense to lift this comment out into a warning about how you can only create one text index per collection? Otherwise I think it's unclear why this "will error".

Also, this comment mentions the "fullplot" field but the code itself uses the "plot" field. Which is correct?

nathan-contino-mongo commented on source/fundamentals/indexes.txt:
```diff
@@ -0,0 +1,253 @@
+
+MongoDB supports queries of geospatial coordinate data using **2dsphere indexes**. With a 2dsphere index, you can query
+the geospatial data for inclusion, intersection, and proximity. For more information on querying geospatial data, see
+:manual:`Geospatial Queries </geospatial-queries/>`.
```
This and other sections include the "For a full explanation" sentence at the very end of the section that already links to this page. The Node driver documentation contains a page on geospatial queries: https://docs.mongodb.com/drivers/node/fundamentals/crud/read-operations/geo
Would it make more sense to link to the Java driver docs equivalent page here, since this link is rendered redundant by the "for a full explanation" link at the end of the section?

nathan-contino-mongo commented on source/fundamentals/indexes.txt:
```diff
@@ -0,0 +1,253 @@
+      }
+   }
+
+The following example uses the ``createIndexes()`` method to create a ``2dsphere`` index on the ``location.geo`` field
```
I noticed that index names are sometimes rendered in monospace text, sometimes plaintext. I think it makes sense for the "text" index above since that's the literal method name you use to create the index, but that's not the case here. Should we use "geo2dsphere" in monospace here instead? Ditto for some of the other index sections regarding monospacing vs. plaintext.

nathan-contino-mongo reviewed (approved):
LGTM with a few nits/suggestions.

nathan-contino-mongo commented on source/fundamentals/indexes.txt:
```diff
@@ -59,6 +59,17 @@ most common index types and provide sample code for creating each index type.
    includes static factory methods to create index specification documents for different MongoDB Index key types.
 
+The following examples use the
+:java-sync-api:`createIndex </com/mongodb/client/MongoCollection.html#createIndex(org.bson.conversions.Bson,com.mongodb.client.model.IndexOptions)>`
```
Noticed something here that we struggle with in our Usage Examples for the Realm SDKs -- normally I like the idea of linking once to a concept at first mention. But for pages where individual sections stand alone as examples, I think it's sometimes useful to link one *per section*, especially since otherwise users might have to hunt through the javadoc to find the ``createIndex()`` method.

terakilobyte replied:
Normally I'd want to link to the method more, but I think there are two mitigating things here:
1. The builders page, which included updating this page with links to it once it is complete, will serve as the primary page for "this is how you do indexes in Java".
2. In all of the IDEs I've tested, this method almost immediately completes and shows signature help.

I'm hesitant to add more items to the page that make it busier with these two things in mind. Thoughts?

nathan-contino-mongo replied:
Seems fair to assume that the builder links and general norm of using an IDE with Java can fill this gap in understanding. LGTM

nathan-contino-mongo commented on source/fundamentals/indexes.txt:
```diff
@@ -117,20 +126,19 @@ See the MongoDB server manual section on :manual:`Compound indexes </core/index-
-**Multikey indexes** are indexes that improve performance for queries that specify ascending or descending indexes on
-fields that contain an array value. You can define a multikey index using the same syntax as a single field or compound
-index.
+**Multikey indexes** are indexes that improve performance for queries that specify an index on a field that contains an
```
Huge improvement here -- this gets to the point with a lot fewer words. Would this be even clearer if we removed "that specify an index", since queries do not have to explicitly specify an index to use it? Or do you think that's cutting too many words?

terakilobyte replied:
I've reworded this slightly, good catch. The emphasis should be on the field queried, which I think it is now.

nathan-contino-mongo commented on source/fundamentals/indexes.txt:
```diff
@@ -149,16 +157,19 @@ Text Indexes
-``movies`` collection in the ``sample_mflix`` database.
+.. warning::
+
+   Collections are limited to one text index. Keep in mind that a text index can cover multiple fields.
```
Good call on the second sentence here!

nathan-contino-mongo commented on source/fundamentals/indexes.txt:
```diff
@@ -205,8 +216,11 @@ database is a GeoJSON Point object that describes the coordinates of the theater
+
+.. note::
+
+   Attemping to create a geospatial index on a field that is covered by a geospatial index will result in an error.
```
Is this a common enough problem to warrant a note? To me, this seems intuitive enough to omit, but maybe I'm missing something? If it helps even a small number of users, I think it's worth keeping, though.

terakilobyte replied:
So, the index story in MongoDB isn't uniform. The other index types that I didn't wrap in a try/catch can be created over and over on the same field, MongoDB will just happily return the name of the existing index.

These other types throw an error. Since I've wrapped the operation in a try/catch I thought it best to add a note.

nathan-contino-mongo replied:
That sounds like a norm-defying footgun to me. LGTM.

nathan-contino-mongo commented on source/fundamentals/indexes.txt:
```diff
@@ -234,17 +248,18 @@ Unique indexes ensure that the indexed fields do not store duplicate values. By
+.. warning::
+
+   If you attempt to perform a write operation that stores a duplicate value that violates the unique index, the MongoDB
+   Java driver will raise a ``DuplicateKeyException``, and MongoDB will throw an error resembling the following:
```
Really liking the "footguns in notes" pattern here that you've established for each section.

nathan-contino-mongo reviewed (approved):
LGTM (again!) with one or two suggestions.

terakilobyte commented:
Ready for review @jyemin @rozza

jyemin commented on source/fundamentals/indexes.txt:
```diff
@@ -0,0 +1,268 @@
+Overview
+--------
+
+Indexes are data structures that support the efficient execution of queries in MongoDB. They contain copies of parts of
```
The use of the phrase `data structures` might confuse people, as typically it is used to refer to in-memory constructs like linked lists or array rather than persistent on-disk representations.  Perhaps instead steal some language from the intro on https://docs.mongodb.com/manual/indexes/?

jyemin commented on source/fundamentals/indexes.txt:
```diff
@@ -0,0 +1,268 @@
+When you execute a query against MongoDB, your query can include three parts:
+
+- query criteria that specify field(s) and value(s) you are looking for
+- options that affect the query's execution (e.g. read concern)
```
This isn't really relevant to indexes.  Consider removing.

terakilobyte replied:
I think it's valuable to frame indexes in how they are used, and this is just a small reminder to readers what a query operation includes. Thoughts?

jyemin replied:
I still find it confusing.  Perhaps if you want to include it, separate it out into a section that makes it clear that it's not relevant to index use.

On a related note, this whole section seems geared towards the find command, though indexes are using for aggregation and updates as well.

jyemin commented on source/fundamentals/indexes.txt:
```diff
@@ -0,0 +1,268 @@
+Query Coverage and Performance
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+When you execute a query against MongoDB, your query can include three parts:
```
Also, sort criteria, which can make use of indexes as well.

terakilobyte replied:
Good point!

jyemin commented on source/fundamentals/indexes.txt:
```diff
@@ -0,0 +1,268 @@
+Query Coverage and Performance
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+When you execute a query against MongoDB, your query can include three parts:
```
I'm concerned that this description is getting ahead of what we say on https://docs.mongodb.com/manual/indexes/#index-use.  And it's incomplete as it doesn't cover, for example, geospatial indexes.

terakilobyte replied:
We'll have the query team lgtm the "common" parts as delimited in the rst

jyemin commented on source/fundamentals/indexes.txt:
```diff
@@ -0,0 +1,268 @@
+- options that affect the query's execution (e.g. read concern)
+- projection criteria to specify the fields MongoDB should return (optional)
+
+When all the fields specified in the query criteria and projection of a query are indexed, MongoDB returns results
```
I'm not sure we should make this promise.  For example, what about sort criteria?  

I think if we're going to include statements like this in driver docs, we need someone from server query to review as well.

jyemin commented on source/fundamentals/indexes.txt:
```diff
@@ -0,0 +1,268 @@
+Single Field Indexes
+++++++++++++++++++++
+
+**Single field indexes** are indexes that improve performance for queries that specify ascending or descending sort
```
This is the first mention of sort criteria.  And it's also not the only or even primary use of a single field index.  For example, the `_id` index.

terakilobyte replied:
We do include mention of the `_id_` index further down in the page, but I added another callout here. I also added a short tip to let readers know indexes help other operations.

jameswahlin replied:
+1 for the sentiment that sort order is not the only use case for a single field index, filtering being the primary example.

jyemin reviewed (changes requested):
I'm concerned about this section.  I found several statements that are either inaccurate or incomplete, and I'm not convinced there aren't more subtleties that I missed entirely. There is a lot of information in here that is describing the behavior of the server, not the driver, and driver engineers are not the right ones to verify its accuracy.

terakilobyte commented:
@cswanson310 Could you give a review to the content in rst wrapped in "common-content-x"?

jameswahlin commented:
> @cswanson310 Could you give a review to the content in rst wrapped in "common-content-x"?

@terakilobyte I can do a review pass for the Query team in Charlie's stead

jameswahlin commented on source/fundamentals/indexes.txt:
```diff
@@ -0,0 +1,340 @@
+- sort criteria to specify the order documents will be returned from MongoDB (optional)
+
+When all the fields specified in the query, projection, and sort are indexed, MongoDB returns results directly from the
+index without scanning any documents in the collection or loading them into memory, also called a **covered query**.
```
I would consider dropping "loading them into memory" as it is not clear what this means? If it is a refers to avoiding an in-memory sort then I think more context will be needed, as the index key order also has to match the sort specification order.

terakilobyte replied:
Thanks, I've removed the terminology. In regards to your point about sorting, do you think the sort order note that follows this paragraph is sufficient in explaining this?

jameswahlin commented on source/fundamentals/indexes.txt:
```diff
@@ -0,0 +1,340 @@
+Compound Indexes
+++++++++++++++++
+
+**Compound indexes** are indexes that improve performance for queries that specify ascending or descending sort order
```
Ditto here on sort just being one use case that indexes exist for.

jameswahlin commented on source/fundamentals/indexes.txt:
```diff
@@ -0,0 +1,340 @@
+.. driver-content-end
+.. common-content-begin
+
+For a full explanation of text search with MongoDB, refer to :manual:`Text Indexes </core/index-text>` in the MongoDB
```
Maybe worth mentioning Atlas Search here? Maybe controversial as it is Atlas only, but could be a better solution for users on Atlas than text indexes.

jameswahlin commented on source/fundamentals/indexes.txt:
```diff
@@ -0,0 +1,340 @@
+server documentation on :manual:`Indexing Strategies </applications/indexes>` and
+:manual:`Data Modeling and Indexes </core/data-model-operations/#data-model-indexes>`.
+
+Index Types
```
Looks like wildcard indexes are missing from this list?

terakilobyte commented on source/fundamentals/indexes.txt:
```diff
@@ -0,0 +1,340 @@
+
+.. tip::
+
+   Indexes are also used by update operations when finding the document(s) to update, and by
```
good point

terakilobyte commented:
@jameswahlin Ready for another look. I've simplified some language and removed emphasis on sorting. I've also listed wildcard indexes in the "operational considerations" section. There's also a callout about Atlas search encouraging users to use it if they are on Atlas.

jameswahlin reviewed (approved):
LGTM

terakilobyte commented:
@jyemin Ready for another pass on the driver specific content.

jyemin commented on source/includes/fundamentals/code-snippets/IndexPage.java:
```diff
@@ -0,0 +1,147 @@
+        cursor.forEach(doc -> System.out.println(doc));
+    }
+
+    private void uniqueIndex() {
```
This method is not called from main, so it's untested.

jyemin commented on source/includes/fundamentals/code-snippets/IndexPage.java:
```diff
@@ -0,0 +1,147 @@
+            IndexOptions indexOptions = new IndexOptions().unique(true);
+            String resultCreateIndex = collection.createIndex(Indexes.descending("theaderId"), indexOptions);
+            System.out.println(String.format("Index created: %s", resultCreateIndex));
+        } catch (DuplicateKeyException e) {
```
Even if I create the index twice in a row, I don't get a DuplicateKeyException (or any exception for that matter).  Is MongoDB just treating the second attempt as a no-op?

terakilobyte replied:
The DuplicateKeyException would be thrown if you tried to create an index with unique constraints on a field that [isn't unique](https://docs.mongodb.com/manual/core/index-unique/#restrictions).

```java
    private void uniqueIndex() {
        System.out.println("unique index");
        collection = database.getCollection("uniqueViolator");

        collection.insertMany(Arrays.asList(new Document("a", 1), new Document("a", 1)));
        // begin unique index
        // this will fail if any duplicate keys (values) exist
        try {
            IndexOptions indexOptions = new IndexOptions().unique(true);
            String resultCreateIndex = collection.createIndex(Indexes.descending("a"), indexOptions);
            System.out.println(String.format("Index created: %s", resultCreateIndex));
        } catch (DuplicateKeyException e) {
            System.out.println("duplicate key, couldn't create index");
        }
        // end unique index
    }
```

terakilobyte commented on source/fundamentals/indexes.txt:
```diff
@@ -0,0 +1,364 @@
+- projection criteria to specify the fields MongoDB should return (optional)
+- sort criteria to specify the order documents will be returned from MongoDB (optional)
+
+When all the fields specified in the query, projection, and sort are indexed, MongoDB returns results directly from the
```
I'll update this, it's an easy clarification to make.

terakilobyte commented:
@jyemin I've pushed some clarifications based on your feedback.

terakilobyte commented on source/includes/fundamentals/code-snippets/IndexPage.java:
```diff
@@ -0,0 +1,147 @@
+            IndexOptions indexOptions = new IndexOptions().unique(true);
+            String resultCreateIndex = collection.createIndex(Indexes.descending("theaderId"), indexOptions);
+            System.out.println(String.format("Index created: %s", resultCreateIndex));
+        } catch (DuplicateKeyException e) {
```
I've also added an imports section to the setup

jyemin reviewed (approved):
LGTM
